### ISSUE TITLE
Unable to set break point

### ISSUE THREAD

**Issue by gunawebs** [rstudio/rstudio#14682]
### System details

    RStudio Edition :Desktop
    RStudio Version : 2024.04.0+735 (2024.04.0+735) 
    OS Version      : 14.4.1 (23E224)
    R Version       : R version 4.3.3 (2024-02-29) -- "Angel Food Cake"

### Steps to reproduce the problem
Create any R code file with a function
Source the file
Add a breakpoint

### Describe the problem in detail
Despite Sourcing the file, breakpoint is NOT placed
Just says "Breakpoints will be activated When this file is sourced"
But the file has ALREADY been sourced.
THIS issue wasn't there before I upgraded the rstudio to the latest version

<img width="421" alt="image" src="https://github.com/rstudio/rstudio/assets/4875731/35746e44-abce-42dc-8a4a-b3846006df2a">

### Describe the behavior you expected
Expected Behaviour: Breakpoint will be placed

- [x ] I have read the guide for [submitting good bug reports](https://github.com/rstudio/rstudio/wiki/Writing-Good-Bug-Reports).
- [ x] I have installed the latest version of RStudio, and confirmed that the issue still persists.
- [ x] If I am reporting an RStudio crash, I have included a [diagnostics report](https://support.posit.co/hc/en-us/articles/200321257-Running-a-Diagnostics-Report).
- [ x] I have done my best to include a minimal, self-contained set of instructions for consistently reproducing the issue.

**Comment by gunawebs** [comment]
IMPORTANT: please note that I have now deleted this version, and installed the earlier version:

Version 2023.12.1+402 (2023.12.1+402)

And breakpoints work JUST FINE

**Comment by gunawebs** [comment]
@ronblum As shared in screenshot above. I was just not able to set breakpoint, in a sourced file. 
And once I installed the prior version, it worked just fine.

**Comment by ronblum** [comment]
@gunawebs I can't reproduce the problem based on the screenshot because part of the code is cut off on the right, it uses objects there were defined elsewhere, and there are at least 219 lines that aren't included.

Does this occur with any sourced file? For example, is it possible to set a breakpoint in the following code?
```
x <- 1
y <- 23
x + y
```

**Comment by sthawinke** [comment]
I encountered the same only in this latest rstudio version, you have to manually resource the function before the breakpoint is set, an any file.

**Comment by gunawebs** [comment]
@ronblum Yes, the issue is with all files. PLs see above comment from @sthawinke  Too!

**Comment by ronblum** [comment]
@sthawinke Can you provide a simple reproducible example? And can you check whether this will cause the problem? I'm trying to find a simple example that I can use to reproduce, because I'm unable to so far. Also, does the problem occur even with this simple example?
```
x <- 1
y <- 23
x + y
```

**Comment by ronblum** [comment]
Thanks! I can reproduce the problem in 2023.03.2+454, 2023.12.1+402, and 2024.07.0-daily+131. So this has always been a requirement. Note this only affects the function; when adding something outside of the function like
```
a <- 8
a
```
the breakpoints will work. This is by design, and only the newly defined functions that are affected.

**Comment by dblodgett-usgs** [comment]
I am seeing basically the same issue in all my packages and code. It's not clear why RStudio 2024.04.1 is not able to debug sourced or properly loaded code. 

To reproduce, 

```r
cat("test <- function(x) {\n  t <- x\n  t\n}",  file = "temp.R")
rstudioapi::documentOpen("temp.R")
source("temp.R")
```

Then I try to set a breakpoint and see:

![image](https://github.com/rstudio/rstudio/assets/1492803/95cad3ad-eee3-44dd-87f8-3dded26cad2b)

I get the same when rebuilding packages except this is the message:
![image](https://github.com/rstudio/rstudio/assets/1492803/c77e949d-a039-43a5-b6ed-d716218063e5)

This is all working under rstudio 2023.12.1+402 and I've tested on both R 4.3.3 and R 4.4.0 in both 2024.04.1+748 and 2023.12.1+402. 

```r
          sysname           release           version 
        "Windows"          "10 x64"     "build 19045" 
         nodename           machine 
           "..."             "x86-64" 
```

**Comment by olivroy** [comment]
In case it is useful, it seems that setting breakpoints is problematic when loading a package that is not installed but is loaded with `load_all()`,

I am seeing the message "Breakpoints will be activated when the package is built is reloaded" message.

Getting an error no package found

Here is the traceback if useful!
```r
Error in loadNamespace(name) : no package named 'EHL' found

10: stop(cond)
9: doWithOneRestart(return(expr), restart)
8: withOneRestart(expr, restarts[[1L]])
7: withRestarts(stop(cond), retry_loadNamespace = function() NULL)
6: loadNamespace(name)
5: getNamespace(ns)
4: asNamespace(packageName)
3: .rs.getEnvironmentOfFunction(functionName, fileName, packageName)
2: .rs.getUntracedFunction(functionName, fileName, packageName)
1: .rs.getFunctionSourceRefs("function_name", "~/rrr/EHL/R/file.R", 
       "EHL")
```

**Comment by ronblum** [comment]
Verified in RStudio Desktop 2024.07.0-daily+244 on Windows 11. Thank you @dblodgett-usgs for the helpful reprex!

**Comment by olivroy** [comment]
This is fixed in the upcoming version of RStudio! You can test it by installing it https://dailies.rstudio.com/

**Comment by rvlenth** [comment]
As suggested, I installed the daily update
```
RStudio 2024.07.0-daily+294 "Cranberry Hibiscus" Daily (eeeb0b44183a826027a25ec0e9482d54de8765c2, 2024-07-08) for windows
Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) RStudio/2024.07.0-daily+294 Chrome/124.0.6367.243 Electron/30.1.2 Safari/537.36, Quarto 1.5.53
```
but I still experience the same issue. **However** I found that I don't have the problem with another package of mine. So I think maybe it has something to do with how we refer to the location of the code. I have it as a long complicated path that contains spaces, and as a shorter one based on a mapped network drive that does not. Both could be problematic, for different reasons. I'll see if I can make sense of that issue and report back.

**Comment by rvlenth** [comment]
I did a test comparing...

1. Starting RStudio by double-clicking the .Rproj file from a window showing the full system path
2. Starting RStudio by double-clicking the .Rproj file from a window showing the mapped network drive

Both gave me the same problem. However from distant past experience, I think it's better to use the full system path, so the rest of this refers to that situation.

Here's another thing I observed. Trying something similar to what I did before...
![image](https://github.com/rstudio/rstudio/assets/26777519/3d40774b-2100-41a7-84f0-6f689a71eb3c)

... but now (with no intervening actions) I clicked in the red bullet in line 124; and darned if the *other* breakpoint is then activated:

![image](https://github.com/rstudio/rstudio/assets/26777519/8d1d3112-cc8c-4307-9bdd-422b362608ca)

Does that provide an additional clue?

**Comment by SergeMerzliakov** [comment]
> IMPORTANT: please note that I have now deleted this version, and installed the earlier version:
> 
> Version 2023.12.1+402 (2023.12.1+402)
> 
> And breakpoints work JUST FINE



> IMPORTANT: please note that I have now deleted this version, and installed the earlier version:
> 
> Version 2023.12.1+402 (2023.12.1+402)
> 
> And breakpoints work JUST FINE

Same here. I had to revert to R 4.3.3 and RStudio and 2023.12.1 +402 on OSX to get debugging to work again. R 4.4.1 does not work with 2023 studio (weird issues with integer types - C/C++ errors)

**Comment by olivroy** [comment]
It may be fixed as part of the latest daily https://github.com/rstudio/rstudio/issues/15072

**Comment by rvlenth** [comment]
@olivroy You said this in reference to the July 8 daily, but it wasn't fixed then and I seriously doubt it is now. I'll wait until I see a release version where they claim that this bug has been addressed.

**Comment by olivroy** [comment]
Sorry I was mistaken the last time. You were right it was not fixed the last time, but please see the newer bug report at #15072. I am pretty sure this has now been fixed in https://dailies.rstudio.com/rstudio/kousa-dogwood/. This release is not set for before October (but pretty sure it is not going to be released before December)

The news bullet is the following in [NEWS.md](https://github.com/rstudio/rstudio/blob/main/NEWS.md)

> Fixed an issue where debugger breakpoints did not function correctly in some cases with R 4.4 (#15072)

**Comment by rvlenth** [comment]
@olivroy Nope, sorry. I downloaded and installed the daily version on my Windows 11 PC

> RStudio 2024.10.0-daily+131 "Kousa Dogwood" Daily (42d4fe22e5e747908d0dff2fb66333fe5b417c06, 2024-09-06) for windows
Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) RStudio/2024.10.0-daily+131 Chrome/126.0.6478.234 Electron/31.4.0 Safari/537.36, Quarto 1.5.57

Then brought up some code in my **emmeans** project. Even after re-installing the package, if I try to set a breakpoint, I get a message that it will be set after I rebuild/reload the package.

## Activation when another breakpoint is de-activated
It also still exhibits the odd behavior I posted earlier on July 8. If I subsequently re-install so that the above breakpoint is activated, then try to set another breakpoint at another line in the same code, I get the rebuild/reload message and an empty circle (so now I have a filled dot for the first breakpoint and an empty circle for the one I just tried to set). But then if I click to de-activate the first breakpoint, it *activates* the second one!

## Spurious highlighting
Another thing I am observing is that when I (unsuccessfully) try to set a breakpoint, often (not always) that line *plus* the line above it (code line or comment) is highlighted so that I now have two lines selected. In fact, sometimes even more than one extra line is highlighted, maybe and especially if they are comments. See illustration below:

![image](https://github.com/user-attachments/assets/cf30434e-9fb8-42c1-9068-48681fc680d5)

If I click again on the red circle, the highlighting reverts to the single line.

## Bracketed vs. un-bracketed function bodies
I have also discovered that I *can* set breakpoints in very short functions where the function body is not enclosed in brackets. I tested this by adding some temporary code and rebuilding. Here is a non-package source file for illustration (where I also tested `=` vs. `<-`):
```
y = 2
z <- 7

dum1 = function(x)
    sqrt(x)

dum2 = function(x) {
    sqrt(x)
}

dum3 <- function(x)
    sqrt(x)

dum4 <- function(x) {
    sqrt(x)
}

```
I *can* set breakpoints in the first two lines and in the bodies of `dum1` and `dum3`; and I *cannot* set breakpoints in `dum2` or `dum4`.

And in my package, I can set breakpoints only in those functions that do not have bracketed bodies.

**Comment by kevinushey** [comment]
Is the package where you're experiencing this behavior available on GitHub / publicly elsewhere? If we can reproduce the issue locally, we'll have a better chance of being able to resolve the issue.

**Comment by rvlenth** [comment]
@kevinushey Yes, **emmeans** is [on GitHub](https://github.com/rvlenth/emmeans). But the first, simplest thing would be to copy the code for bracketed/unbracketed behavior into a new R source code window on the Windows version of RSruydio,, and see if you have the same experience.

**Comment by rvlenth** [comment]
I also have several projects for different packages, and I experience this issue with all of them.

**Comment by kevinushey** [comment]
> Bracketed vs. un-bracketed function bodies

In this example, the breakpoint is actually being set at the top-level; that is, one the code that defines the function, as opposed to within the body of the function itself. That is, the breakpoint is likely not doing what you're expecting / hoping it to do.

In any event, some more fixes have come to the latest daily builds at https://dailies.rstudio.com/; some testing would be appreciated.

**Comment by rvlenth** [comment]
Okay! I installed the following daily version:

> RStudio 2024.11.0-daily+258 "Kousa Dogwood" Daily (2175496934a54e21979461a1c67eb13ad5048cf2, 2024-10-02) for windows
Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) RStudio/2024.11.0-daily+258 Chrome/126.0.6478.234 Electron/31.5.0 Safari/537.36, Quarto 1.5.57

... and so far, I **have** been successful at setting breakpoints! I have done a test where I actually used them for debugging a call, and the breakpoints have worked correctly.
